### PR TITLE
Integrating Rust code into ClickHouse

### PR DESCRIPTION
**Pull request by BoloniniD**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add Rust code support into ClickHouse with BLAKE3 hash-function library as an example.

> Does not support shared builds for now, as corrosion-rs tries to use libraries unknown to linker while building. Currently Rust always builds as static libraries.
> To update a Rust library, edit its Cargo.toml. Change the version of your library to newer one, then check if library builds with CH. If something does not work, check changed functions and methods if the lib and update shim methods in lib.rs accordingly.

### CODE REVIEW

File: contrib/CMakeLists.txt [hunk 1]
```diff
@@ -91,6 +91,7 @@ add_contrib (protobuf-cmake protobuf)
 add_contrib (openldap-cmake openldap)
 add_contrib (grpc-cmake grpc)
 add_contrib (msgpack-c-cmake msgpack-c)
+add_contrib (BLAKE3)
```
Comment (vitlibar) [inline contrib/CMakeLists.txt:94]: It can be useful to keep the possibility to build ClickHouse without Rust and its components.

Comment (vitlibar) [inline contrib/CMakeLists.txt:94]: `contrib/blake3` should not refer to your personal repository.
It's better to make it refer to the official blake3 repository, and then provide a custom `CMakeLists.txt` file in the separate directory blake3-cmake.

**Comment by alexey-milovidov** [comment]
To merge this pull request, we need to fix the issue with the linking order, which is leading to exposing the dependency on the unwanted symbols from Glibc (as shown by "Compatibility check — glibc check failed")

**Comment by alexey-milovidov** [comment]
All green :green_apple:

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -588,6 +592,31 @@ struct ImplXxHash64
...
+        #if defined(MEMORY_SANITIZER)
+            auto err_msg = blake3_apply_shim_msan_compat(begin, size, out_char_data);
+            __msan_unpoison(out_char_data, length);
+        #else
+            auto err_msg = blake3_apply_shim(begin, size, out_char_data);
+        #endif
+        if (err_msg != nullptr)
+        {
+            throw Exception("Function returned error message: " + std::string(err_msg), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            blake3_free_char_pointer(err_msg);
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsHashing.h:613]: .

File: docker/test/performance-comparison/Dockerfile [hunk 3]
```diff
@@ -35,7 +35,9 @@ RUN apt-get update \
             tzdata \
             vim \
             wget \
-    && pip3 --no-cache-dir install 'clickhouse-driver==0.2.1' scipy \
+            rustc \
+            cargo \
+    && pip3 --no-cache-dir install 'clickhouse-driver==0.2.1' numpy==1.14.5 scipy \
```
Comment (alexey-milovidov) [inline docker/test/performance-comparison/Dockerfile:40]: This modification `numpy==1.14.5` looks unrelated but ok.

File: docker/packager/binary/build.sh [hunk 4]
```diff
@@ -76,6 +76,8 @@ else
     cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA "-DCMAKE_BUILD_TYPE=$BUILD_TYPE" "-DSANITIZE=$SANITIZER" -DENABLE_CHECK_HEAVY_BUILDS=1 "${CMAKE_FLAGS[@]}" ..
 fi
 
+
```
Comment (alexey-milovidov) [inline docker/packager/binary/build.sh:79]: Unrelated.

File: contrib/BLAKE3/build_rust_lib.cmake [hunk 5]
```diff
@@ -0,0 +1,106 @@
+function(build_cargo target_name project_dir)
```
Comment (vitlibar) [inline contrib/BLAKE3/build_rust_lib.cmake:1]: It seems [corrosion](https://github.com/corrosion-rs/corrosion) might be able to provide the same functionality out of the box. Check it out: https://github.com/ClickHouse/ClickHouse/pull/41225

File: rust/BLAKE3/include/blake3.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#include <cstdarg>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <ostream>
```
Comment (vitlibar) [inline rust/BLAKE3/include/blake3.h:5]: cstdint seems to be enough

And please add usual C guards:

```
#ifndef BLAKE3_H
#define BLAKE3_H

...

#endif /* BLAKE3_H */
```

File: src/Functions/FunctionsHashing.cpp [hunk 7]
```diff
@@ -41,5 +41,22 @@ REGISTER_FUNCTION(Hashing)
...
+Calculates BLAKE3 hash string and returns the resulting set of bytes as FixedString.
+This cryptographic hash-function is integrated into ClickHouse with BLAKE3 Rust library.
+The function is rather fast and shows approximately two times faster performance compared to SHA-2, while generating hashes of the same length as SHA-256.
+It returns a BLAKE3 hash as a byte array with type FixedString(32).
+)",
+        Documentation::Examples{
+            {"hash", "SELECT hex(BLAKE3('ABC'))"}},
+        Documentation::Categories{"Hash"}
+    },
+    FunctionFactory::CaseInsensitive);
```
Comment (vitlibar) [inline src/Functions/FunctionsHashing.cpp:59]: Why should it be case insensitive?

File: contrib/corrosion-cmake/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,2 @@
+list (APPEND CMAKE_MODULE_PATH "${ClickHouse_SOURCE_DIR}/contrib/corrosion/cmake")
```
Comment (vitlibar) [inline contrib/corrosion-cmake/CMakeLists.txt:1]:
```
if NOT ENABLE_LIBRARIES
  option(ENABLE_RUST "Enable rust" FALSE)
else()
  list (APPEND CMAKE_MODULE_PATH "${ClickHouse_SOURCE_DIR}/contrib/corrosion/cmake")
  find_package(Rust)
  option(ENABLE_RUST "Enable rust" $Rust_FOUND)
endif()

if(NOT ENABLE_RUST)
  message(STATUS "Not using rust")
  return()
endif()

# Define function corrosion_import_crate()
include ("${ClickHouse_SOURCE_DIR}/contrib/corrosion/cmake/Corrosion.cmake")
```

File: base/glibc-compatibility/spawn.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+int posix_spawnattr_setsigdefault(posix_spawnattr_t *__restrict, const sigset_t *__restrict);
+int posix_spawnattr_getsigdefault(const posix_spawnattr_t *__restrict, sigset_t *__restrict);
+
+int posix_spawnattr_setschedparam(posix_spawnattr_t *__restrict, const struct sched_param *__restrict);
+int posix_spawnattr_getschedparam(const posix_spawnattr_t *__restrict, struct sched_param *__restrict);
+int posix_spawnattr_setschedpolicy(posix_spawnattr_t *, int);
+int posix_spawnattr_getschedpolicy(const posix_spawnattr_t *__restrict, int *__restrict);
+
+int posix_spawn_file_actions_init(posix_spawn_file_actions_t *);
+int posix_spawn_file_actions_destroy(posix_spawn_file_actions_t *);
```
Comment (vitlibar) [inline base/glibc-compatibility/spawn.h:66]: Who calls these functions?

File: contrib/CMakeLists.txt [hunk 10]
```diff
@@ -92,6 +92,10 @@ add_contrib (openldap-cmake openldap)
 add_contrib (grpc-cmake grpc)
 add_contrib (msgpack-c-cmake msgpack-c)
 
+#if (ENABLE_RUST)
```
Comment (vitlibar) [inline contrib/CMakeLists.txt:95]: #if (ENABLE_RUST) is not necessary

Comment (BoloniniD) [reply contrib/CMakeLists.txt:95]: Flag ENABLE_RUST is used to disable Rust compilation completely. Corrosion is not needed if Rust is disabled, so we exclude it from build.

File: contrib/corrosion-cmake/CMakeLists.txt (hunk 8 above)
Comment (BoloniniD) [reply contrib/corrosion-cmake/CMakeLists.txt:1]: Looks nice, replaced previous CMakeLists.txt code with yours

File: base/glibc-compatibility/spawn.h (hunk 9 above)
Comment (BoloniniD) [reply base/glibc-compatibility/spawn.h]: Removed unused functions in new commit. posix_spawn is being called in glibc_compatibility.c

File: contrib/corrosion-cmake/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  option(ENABLE_RUST "Enable rust" FALSE)
+elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "aarch64"))
+  option(ENABLE_RUST FALSE)
+  message(STATUS "Rust is not available on aarch64-apple-darwin")
+  return()
+else()
+  list (APPEND CMAKE_MODULE_PATH "${ClickHouse_SOURCE_DIR}/contrib/corrosion/cmake")
+  find_package(Rust)
+  option(ENABLE_RUST "Enable rust" ${Rust_FOUND})
+endif()
```
Comment (vitlibar) [inline contrib/corrosion-cmake/CMakeLists.txt:11]:
```suggestion
if (NOT ENABLE_LIBRARIES)
  set(DEFAULT_ENABLE_RUST FALSE)
elseif((CMAKE_TOOLCHAIN_FILE MATCHES "darwin") AND (CMAKE_TOOLCHAIN_FILE MATCHES "aarch64"))
  message(STATUS "Rust is not available on aarch64-apple-darwin")
  set(DEFAULT_ENABLE_RUST FALSE)
else()
  list (APPEND CMAKE_MODULE_PATH "${ClickHouse_SOURCE_DIR}/contrib/corrosion/cmake")
  find_package(Rust)
  set(DEFAULT_ENABLE_RUST ${Rust_FOUND})
endif()

option(ENABLE_RUST "Enable rust" ${DEFAULT_ENABLE_RUST})
```

seems a bit better because this way `option(ENABLE RUST ...)` is written only once

File: docker/test/fasttest/run.sh [hunk 12]
```diff
@@ -158,6 +158,7 @@ function run_cmake
         "-DENABLE_NURAFT=1"
         "-DENABLE_JEMALLOC=1"
         "-DENABLE_REPLXX=1"
+        "-DENABLE_RUST=0"
```
Comment (vitlibar) [inline docker/test/fasttest/run.sh:161]: Why do you need this? Just don't install Rust and it will set `ENABLE_RUST=0` automatically, won't it?

File: rust/BLAKE3/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+version = "0.1.0"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+blake3 = "1.2.*"
+libc = "0.2.*"
+
+[lib]
+crate-type = ["staticlib"]
```
[the remark this answers is not in the document]
Comment (BoloniniD) [reply rust/BLAKE3/Cargo.toml:12]: With `cdylib` enabled, corrosion adds some standard libraries even if NO_STD flag is used. This can cause errors during Rust library linking, so I decided to leave only `staticlib` for now.

File: docker/test/fasttest/run.sh (hunk 12 above)
Comment (BoloniniD) [reply docker/test/fasttest/run.sh]: Removed the flag, everything works without it in fasttest

File: src/Functions/FunctionsHashing.cpp [hunk 14]
```diff
@@ -41,5 +41,22 @@ REGISTER_FUNCTION(Hashing)
     factory.registerFunction<FunctionXxHash64>();
 
     factory.registerFunction<FunctionWyHash64>();
+
+
+#if USE_BLAKE3
```
Comment (vitlibar) [inline src/Functions/FunctionsHashing.cpp:46]: I think it's better to register the function anyway but throw an exception `SUPPORT_IS_DISABLED` with a description why it's disabled without too much details (i.e. what we lack of: rust, blake3 is disabled or maybe it's aarch64). Because if you don't register the function an user will see an exception like "Function is not found" which is not very helpful in this case.

Comment (BoloniniD) [reply src/Functions/FunctionsHashing.cpp:46]: Opened #42073 to implement this
